pkg/machines/v1alpha1/machineset_types.go

```diff
@@ -0,0 +1,167 @@
+	// +optional
+	AvailableReplicas int32 `json:"availableReplicas,omitempty"`
+
+	// ObservedGeneration reflects the generation of the most recently observed MachineSet.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// In the event that there is a terminal problem reconciling the
+	// replicas, both ErrorReason and ErrorMessage will be set. ErrorReason
+	// will be populated with a succinct value suitable for machine
+	// interpretation, while ErrorMessage will contain a more verbose
+	// string suitable for logging and human consumption.
+	//
+	// These fields should not be set for transitive errors that a
+	// controller faces that are expected to be fixed automatically over
+	// time (like service outages), but instead indicate that something is
+	// fundamentally wrong with the MachineTemplates's spec or the configuration of
+	// the machine controller, and that manual intervention is required. Examples
+	// of terminal errors would be invalid combinations of settings in the
+	// spec, values that are unsupported by the machine controller, or the
+	// responsible machine controller itself being critically misconfigured.
+	//
+	// Any transient errors that occur during the reconcilation of Machines
+	// can be added as events to the MachineSet object and/or logged in the
+	// controller's output.
+	// +optional
+	ErrorReason *common.MachineSetStatusError `json:"errorReason,omitempty"`
+	// +optional
+	ErrorMessage *string `json:"errorMessage,omitempty"`
+}
+
+// Validate checks that an instance of MachineSet is well formed
+//func (MachineSetStrategy) Validate(ctx request.Context, obj runtime.Object) field.ErrorList {
+//	machineSet := obj.(*MachineSet)
+//	errors := field.ErrorList{}
+//
+//	// validate spec.selector and spec.template.labels
+//	fldPath := field.NewPath("spec")
+//	errors = append(errors, metav1validation.ValidateLabelSelector(&machineSet.Spec.Selector, fldPath.Child("selector"))...)
+//	if len(machineSet.Spec.Selector.MatchLabels)+len(machineSet.Spec.Selector.MatchExpressions) == 0 {
+//		errors = append(errors, field.Invalid(fldPath.Child("selector"), machineSet.Spec.Selector, "empty selector is not valid for MachineSet."))
+//	}
+//	selector, err := metav1.LabelSelectorAsSelector(&machineSet.Spec.Selector)
+//	if err != nil {
+//		errors = append(errors, field.Invalid(fldPath.Child("selector"), machineSet.Spec.Selector, "invalid label selector."))
+//	} else {
+//		labels := labels.Set(machineSet.Spec.Template.Labels)
+//		if !selector.Matches(labels) {
+//			errors = append(errors, field.Invalid(fldPath.Child("template", "metadata", "labels"), machineSet.Spec.Template.Labels, "`selector` does not match template `labels`"))
+//		}
+//	}
+//
+//	return errors
+//}
+//
+//// DefaultingFunction sets default MachineSet field values
+//func (MachineSetSchemeFns) DefaultingFunction(o interface{}) {
```

Comment:
Same here, `MachineSetSchemeFns` also needs apiserver builder which is why I commented it out

Reply:
https://github.com/kubermatic/machine-controller/pull/287/files#r205680919

Reply:
The defaulting _could_ actually be used within a webhook